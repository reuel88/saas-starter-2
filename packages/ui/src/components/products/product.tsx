import { FC } from "react";
import { Card, CardFooter, CardHeader, CardTitle } from "../card";
import { Button } from "../button";
import { Carousel, CarouselContent, CarouselItem } from "../ui/carousel";
import { ProductRating } from "./product-rating";

export type ProductProps = {
  title: string;
  price: number;
  rating: number;
  images?: { imageSrc: string }[];
};

export const Product: FC<ProductProps> = ({ title, price, rating, images }) => {
  return (
    <Card as="article">
      <Carousel>
        <CarouselContent>
          {images?.map((image, index) => (
            <CarouselItem key={index}>
              <img
                src={image.imageSrc}
                alt={title}
                className="aspect-square rounded-t-lg object-cover"
              />
            </CarouselItem>
          ))}
        </CarouselContent>
      </Carousel>

      <CardHeader className="space-y-0">
        <CardTitle className="line-clamp-2 text-base font-normal text-muted-foreground">
          {title}
        </CardTitle>
        <div className="flex items-center justify-between">
          <p className="text-lg font-bold">
            {new Intl.NumberFormat("en-AU", {
              style: "currency",
              currency: "AUD",
              minimumFractionDigits: 0,
            }).format(price)}
          </p>
          <div>
            <ProductRating rating={rating} />
          </div>
        </div>
      </CardHeader>
      <CardFooter as="footer">
        <Button type="button" className="w-full">
          Add to Cart
        </Button>
      </CardFooter>
    </Card>
  );
};
