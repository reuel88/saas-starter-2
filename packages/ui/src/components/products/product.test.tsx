import { render, screen } from "@testing-library/react";
import { Product, ProductProps } from "./product";
import { vi } from "vitest";
import { ReactNode } from "react";

vi.mock("../ui/carousel", async (importOriginal) => {
  return {
    ...(await importOriginal<typeof import("../ui/carousel")>()),
    CarouselContent: ({ children }: { children: ReactNode }) => {
      return <div>{children}</div>;
    },
  };
});

describe("Product", () => {
  it("should render a Product component inside, displaying title, price and rating", () => {
    // Arrange
    const product: ProductProps = {
      title: "Example Product",
      price: 29.99,
      rating: 4.5,
      images: [
        { imageSrc: "image1.jpg" },
        { imageSrc: "image2.jpg" },
        { imageSrc: "image3.jpg" },
      ],
    };

    // Act
    const { baseElement } = render(<Product {...product} />);
    const cardHeaderElement = screen.getByRole("heading", { level: 3 });
    const cardTitleElement = screen.getByText(product.title);
    const priceElement = screen.getByText("$29.99");
    const productRatingElement = screen.getByText(`(${product.rating})`);

    // Assert
    expect(cardHeaderElement).toBeInTheDocument();
    expect(cardTitleElement).toBeInTheDocument();
    expect(priceElement).toBeInTheDocument();
    expect(productRatingElement).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });
});
