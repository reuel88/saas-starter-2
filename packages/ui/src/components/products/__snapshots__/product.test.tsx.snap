// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product > should render a Product component inside, displaying title, price and rating 1`] = `
<body>
  <div>
    <article
      class="rounded-lg border bg-card text-card-foreground shadow-sm"
    >
      <div
        aria-roledescription="carousel"
        class="relative"
        role="region"
      >
        <div>
          <div
            aria-roledescription="slide"
            class="min-w-0 shrink-0 grow-0 basis-full pl-4"
            role="group"
          >
            <img
              alt="Example Product"
              class="aspect-square rounded-t-lg object-cover"
              src="image1.jpg"
            />
          </div>
          <div
            aria-roledescription="slide"
            class="min-w-0 shrink-0 grow-0 basis-full pl-4"
            role="group"
          >
            <img
              alt="Example Product"
              class="aspect-square rounded-t-lg object-cover"
              src="image2.jpg"
            />
          </div>
          <div
            aria-roledescription="slide"
            class="min-w-0 shrink-0 grow-0 basis-full pl-4"
            role="group"
          >
            <img
              alt="Example Product"
              class="aspect-square rounded-t-lg object-cover"
              src="image3.jpg"
            />
          </div>
        </div>
      </div>
      <div
        class="flex flex-col p-6 space-y-0"
      >
        <h3
          class="tracking-tight line-clamp-2 text-base font-normal text-muted-foreground"
        >
          Example Product
        </h3>
        <div
          class="flex items-center justify-between"
        >
          <p
            class="text-lg font-bold"
          >
            $29.99
          </p>
          <div>
            <div
              class="flex"
            >
              <svg
                class="h-5 w-5 text-hue-color"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="h-5 w-5 text-hue-color"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="h-5 w-5 text-hue-color"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="h-5 w-5 text-hue-color"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="h-5 w-5 text-hue-color"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <span
                class="text-muted-foreground"
              >
                (
                4.5
                )
              </span>
            </div>
          </div>
        </div>
      </div>
      <footer
        class="flex items-center p-6 pt-0"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full"
          type="button"
        >
          Add to Cart
        </button>
      </footer>
    </article>
  </div>
</body>
`;
